Add documentation on migrating node.js

yucheng11122017 opened the pull request:
**What is the purpose of this pull request?**

- [X] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

In response to [comment](https://github.com/MarkBind/markbind/pull/2233#issuecomment-1486060765) in #2233
<!--
  If this pull request is addressing an issue, link to the issue: #xxx

  If this pull request completely addresses an issue, use one of the closing keywords: "Fixes #xxx" or "Resolves #xxx"
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Added page on devGuide on how to migrate Node.js

**Anything you'd like to highlight/discuss:**


**Testing instructions:**

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Add documentation on migrating node.js

<!--
  See this link for more info on how to write a good commit message:
  https://se-education.org/guides/conventions/git.html

  As best as possible, write a succinct commit title in 50 characters

|---------This is the width of 50 chars----------|
|-----------This is the width of 72 chars for your reference-----------|
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [X] Linked all related issues
- [X] No unrelated changes <!-- It's tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

jovyntls commented on docs/devGuide/development/migratingNodeJs.md:
```diff
@@ -0,0 +1,97 @@
+
+<div class="lead">
+
+Node.js versions have to upgraded periodically before the version being used reaches its' [end of life](https://endoflife.date/nodejs).
```
```suggestion
Node.js versions have to upgraded periodically before the version being used reaches its [end of life](https://endoflife.date/nodejs).
```

jovyntls commented on docs/devGuide/development/migratingNodeJs.md:
```diff
@@ -0,0 +1,97 @@
+To specify the npm version add an environment variable `NPM_VERSION` with the correct version number. 
+</box>
+
+   - Deployment to Github pages
```
The indentation here is a bit off - I think it's rendering as one level less than intended

jovyntls commented on docs/devGuide/development/migratingNodeJs.md:
```diff
@@ -0,0 +1,97 @@
+
+<box type="tip" seamless>
+
+Don't forget to update the version numbers in the example config files in [Deploying the Site]({{baseUrl}}/userGuide/deployingTheSite.html)
```
```suggestion
Don't forget to update the version numbers in the example config files in [Deploying the Site]({{baseUrl}}/userGuide/deployingTheSite.html)!
```

jovyntls reviewed (changes requested):
The guide looks comprehensive to me! Thank you @yucheng11122017 🎉 

Minor nits: can we standardise the capitalisation of keywords like `npm`, `node`?

yucheng11122017 commented:
> The guide looks comprehensive to me! Thank you @yucheng11122017 🎉
> 
> Minor nits: can we standardise the capitalisation of keywords like `npm`, `node`?

Thanks for the review @jovyntls!  I fixed the nits accordingly. The indentation of the box is also now inline with the bullet points. 

I changed all npm to 'npm' and node to 'Node.js'. Hope thats ok!

EltonGohJH reviewed (approved):
LGTM

jovyntls reviewed (changes requested):
<img width="924" alt="Untitled" src="https://user-images.githubusercontent.com/61113575/229684157-45961db3-fc43-43be-bc71-02ef300afd84.png">

Is it possible to not have this newline between a numbered point and its sub points?

yucheng11122017 commented:
> Thanks for the fixes @yucheng11122017! Is it possible to not have this newline between a numbered point and its sub points?
> 
> <img alt="Untitled" width="924" src="https://user-images.githubusercontent.com/61113575/229684157-45961db3-fc43-43be-bc71-02ef300afd84.png">

I tried to remove that line but it seems that this is margin that is automatically added in the bullets. I tried to follow these two posts ([Link 1](https://stackoverflow.com/questions/50163820/how-to-write-an-ordered-list-which-contains-nested-unordered-lists-github-markd) and [Link 2](https://stackoverflow.com/questions/45113083/how-to-remove-extra-line-space-on-github-markdown-bullets-lists)) but seems like its not possible. 

Would it be better if I added spacing between the sub headings similar to [Migrating to TypeScript](https://markbind.org/devdocs/devGuide/development/migratingToTypeScript.html#:~:text=Bundled%20alongside%20their%20implementation%2C%20you,D%20%40types/%3Clib%2Dname%3E.)?

![image](https://user-images.githubusercontent.com/77204346/229731994-114b250a-7928-4df8-bb87-975297744176.png)

damithc commented:
> I tried to remove that line but it seems that this is margin that is automatically added in the bullets. I tried to follow these two posts ([Link 1](https://stackoverflow.com/questions/50163820/how-to-write-an-ordered-list-which-contains-nested-unordered-lists-github-markd) and [Link 2](https://stackoverflow.com/questions/45113083/how-to-remove-extra-line-space-on-github-markdown-bullets-lists)) but doesn't work.

Yes, this is a MarkBind quirk I encounter often when writing bullet lists.

jovyntls commented:
Using two indentations instead of one indentation per bullet point should fix this issue! Leaving a minimal reproduction here in case this is useful for  @damithc :

```md
### Migration steps 

1. first numbered point
    - xxx (this has two indentation levels instead of one)
    - xxx
1. second numbered point
    - xxx
    - xxx
1. third numbered point
```
output:
<img width="285" alt="image" src="https://user-images.githubusercontent.com/61113575/229961210-5292ad08-7d63-4e42-8108-b5f043b9c586.png">

Maybe we should document this somewhere 🤔 

As for this specific example, it works using two indentations per bullet point! The `<box>` tags complicate things a little but after playing around with the spacing this seems to work for me:

```md
### Migration steps 

1. Refactor any deprecated syntax
    - Go to the [Node.js changelog](https://nodejs.org/en/blog/release) of the new version
    - Go through list of deprecated syntax and check if it is being used in Markbind
    - Replace any deprecated syntax
2. Check that all user-facing functionalities are working 
    - A quick way to do this is to go to the [Reader Facing Features in the User Guide]({{baseUrl}}/userGuide/readerFacingFeatures.html) and check that these features are still working as expected.
3. Check that there are no issues with development setup 
    - Set up the development environment by running through the steps in [Setting Up]({{baseUrl}}/devdocs/devGuide/development/settingUp.html) to ensure there are no problems
4. Update GitHubActions 
    - Go to [Markbind/markbind-action](https://github.com/MarkBind/markbind-action) and update the Node.js version numbers
        - See [Update node version from 14 to 16 PR](https://github.com/MarkBind/markbind-action/pull/8/files) for an example
    - Test there are no issues with workflows
        - Testing instructions located here: [markbind-action]({{baseUrl}}/devdocs/devGuide/githubActions/markbindAction.html) and [markbind-reusable-workflows]({{baseUrl}}/devdocs/devGuide/githubActions/markbindReusableWorkflows.html) 
        <box type="info" seamless header="If a different npm version is needed"> 
        Install correct version of npm in `action.yml` and `fork-build.yml`. Refer to [Update node version from 14 to 16 PR](https://github.com/MarkBind/markbind-action/pull/8/files) to see where npm install should be run. </box>
5. Check deployment to Netlify/other platforms
    - Deployment to Netlify 
        - Follow steps in [Deploying to Netlify]({{baseUrl}}/userGuide/deployingTheSite.html#deploying-to-netlify) but change the `NODE_VERSION` value accordingly. Check there are no issues with deployment and deployed site is as expected.
        - Markbind has two repos [init-minimal-netlify](https://github.com/MarkBind/init-minimal-netlify) and [init-typical-netlify](https://github.com/MarkBind/init-typical-netlify) which allows deployment to Netlify by using a config file. Update the config file `netlify.toml` with the correct Node.js version and check that deployment using button in `README` works as expected.
        <box type="info" seamless header="If a different npm version is needed">
        To specify the npm version add an environment variable `NPM_VERSION` with the correct version number. </box>
    - Deployment to Github pages
        - Using the `markbind deploy command`
            - Build site using correct Node.js version 
            - Follow steps in [Using the `markbind deploy` command]({{baseUrl}}/userGuide/deployingTheSite.html#deploying-to-github-pages) and check there are no issues with deployment or with the site. 
        - Using CI platforms
            - Follow steps in [Using CI Platforms]({{baseUrl}}/userGuide/deployingTheSite.html#using-ci-platforms) but update the config files for the various CI Platforms to use the correct Node.js version. Try deploying and ensure there are no problems with deployment. 
            <box type="info" seamless header="If a different npm version is needed">
            Install the correct npm version before running npm commands. </box>
6. Update documentation
    - Update Node.js and npm version in documentation. See [Update to use Node 16](https://github.com/MarkBind/markbind/pull/2233/files#diff-0f8e38868f41667abec6adacbb5131fbd6999c4913fc43e3429390b744f7a1f3) as an example. <box type="tip" seamless>
Don't forget to update the version numbers in the example config files in [Deploying the Site]({{baseUrl}}/userGuide/deployingTheSite.html)!
</box>
```
<img width="992" alt="image" src="https://user-images.githubusercontent.com/61113575/229961500-7229172f-978c-4b59-93ad-4f60a34d1afd.png">

jovyntls commented on docs/devGuide/development/migratingNodeJs.md:
```diff
@@ -0,0 +1,95 @@
+<div class="lead">
+
+Node.js versions have to upgraded periodically before the version being used reaches its [end of life](https://endoflife.date/nodejs).
+This page outlines the steps of migrating to a higher version of Node.js
```
```suggestion
This page outlines the steps of migrating to a higher version of Node.js.
```

jovyntls commented on docs/devGuide/development/migratingNodeJs.md:
```diff
@@ -0,0 +1,95 @@
+
+Read more about Node.js release lines [here](https://nodesource.com/blog/understanding-how-node-js-release-lines-work/).
+
+npm version will be upgraded accordingly to the Node.js version. Hence, do check that this upgrade does not cause any issues.
```
Just checking, what does "accordingly to" refer to? Is it automatically upgraded or should I take any action?

```suggestion
The npm version will be upgraded accordingly to the Node.js version. Hence, do check that this upgrade does not cause any issues.
```

yucheng11122017 replied:
It should be upgraded automatically! I added further clarification in 
7eff420

jovyntls commented on docs/devGuide/development/migratingNodeJs.md:
```diff
@@ -0,0 +1,95 @@
+
+1. Refactor any deprecated syntax
+   - Go to the [Node.js changelog](https://nodejs.org/en/blog/release) of the new version
+   - Go through list of deprecated syntax and check if it is being used in Markbind
```
```suggestion
   - Go through the list of deprecated syntax and check if it is being used in MarkBind
```

Let's standardise to use "MarkBind" (instead of "Markbind") throughout! Same for the other occurrences in this file :)

jovyntls reviewed (changes requested):
A few minor consistency nits (and a clarification)!

damithc commented:
> Using four spaces per indent should fix this issue! Leaving a minimal reproduction here in case this is useful for @damithc future uses:

@jovyntls Thanks for the tip. I tried it and it looks like a blank line anywhere in the middle the list causes a blank line to appear _below_ every bullet point.


```
1. first point
    - some point
    - another point
1. second point
1. third point

--------------------

1. first point
    - some point
    - another point

1. second point
1. third point
```

![image](https://user-images.githubusercontent.com/1673303/229989628-4a8eed7c-eda5-40e6-b8ca-076add64199a.png)

yucheng11122017 commented:
Thanks for helping fix the indentation problems @jovyntls! I think it didn't work because of what @damithc mentioned 

> > Using four spaces per indent should fix this issue! Leaving a minimal reproduction here in case this is useful for @damithc future uses:
> 
> @jovyntls Thanks for the tip. I tried it and it looks like a blank line anywhere in the middle the list causes a blank line to appear _below_ every bullet point.

Also fixed the nits in the documentation

jovyntls reviewed (approved):
LGTM!
